[HU JIAJUN] iP

HolmesJJ opened the pull request:
### HomlesJJ の Duke

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

**HomlesJJ の Duke** frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download [CLI](https://github.com/HolmesJJ/ip/releases/tag/v0.2) or [GUI](https://github.com/HolmesJJ/ip/releases/tag/v0.3) version.
2. Open terminal.
3. Use `cd` command to navigate to the directory of jar file.
4. Enter `java -jar duke.jar` (Prerequisites: [jdk](https://www.oracle.com/sg/java/technologies/javase-jdk11-downloads.html) installed).
5. add your tasks.
6. let it manage your tasks for you :wink:

And it is **FREE!**

Features:

- [x] Managing Todo Tasks
- [x] Managing Event Tasks
- [x] Managing Deadline Tasks

---

If you Java programmer, you can use it to practice Java too. Here's the main method:

For CLI
```
public class Duke implements Message {
    // ...
    public static void main(String[] args) {
        new Duke().run();
    }
}
```

For GUI
```
public class Main {
    public static void main(String[] args) {
        Application.launch(DukeApp.class, args);
    }
}
```

You can also find the source code at [here](https://github.com/HolmesJJ/ip). 🤩

Commit:
Add Gradle support

HolmesJJ committed:
Implement Level-1

HolmesJJ committed:
Implement Level-2

HolmesJJ committed:
Implement Level-3

HolmesJJ committed:
Implement Level-4

HolmesJJ committed:
Implement A-TextUiTesting

HolmesJJ committed:
Implement Level-5

HolmesJJ committed:
Implement Level-6

HolmesJJ committed:
Add more exception types

HolmesJJ committed:
Merge branch 'branch-Level-5'

HolmesJJ committed:
Fix horizontal line and exception bugs

HolmesJJ committed:
Implement Level-7

HolmesJJ committed:
Merge branch 'branch-Level-7'

HolmesJJ committed:
Implement Level-8

HolmesJJ committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/task/Deadline.java
#	src/main/java/task/Event.java

HolmesJJ committed:
Implement A-MoreOOP

HolmesJJ committed:
Implement A-JUnit

HolmesJJ committed:
Implement A-Jar

HolmesJJ committed:
Implement checkstyle

HolmesJJ committed:
Use stream instead of loop

HolmesJJ committed:
Implement Level-9

HolmesJJ committed:
Merge branch 'branch-Level-9'

HolmesJJ committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/util/CommandUtils.java

HolmesJJ committed:
Ignore Jar build files

HolmesJJ committed:
Merge branch 'branch-A-CodingStandard'

HolmesJJ committed:
Update test method name and remove incorrect test method

HolmesJJ committed:
Merge branch 'branch-A-JUnit'

# Conflicts:
#	src/test/java/task/DeadlineTest.java
#	src/test/java/task/EventDateTimeTest.java
#	src/test/java/task/EventTest.java
#	src/test/java/task/TaskManagerTest.java
#	src/test/java/util/CommandUtilsTest.java

HolmesJJ committed:
Update format of Javadoc comments

HolmesJJ committed:
Update class name and variable name

HolmesJJ committed:
Merge branch 'branch-A-CodingStandard'

HolmesJJ committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/storage/Storage.java
#	src/main/java/task/TaskList.java
#	src/main/java/util/Parser.java

HolmesJJ committed:
Move all the classes into duke package

HolmesJJ committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

HolmesJJ committed:
Update gradle setting

HolmesJJ committed:
Update coding standard based on checkstyle

HolmesJJ committed:
Implement Level-10

HolmesJJ committed:
Update code format and comment

HolmesJJ committed:
Remove the unnecessary file

HolmesJJ committed:
Update build.gradle to release GUI version

HolmesJJ committed:
Implement Assertions

Implementing Assertions for the important assumptions that should hold in the code.

Let's
1. Enable assertions at build.gradle.
2. Add assert keyword for the String split length.
    * TaskList: The saved task should have the correct format which has be defined when the task to be saved. So when the task is loaded from file, its format is correct by default, means it should be able to be split to the required length and hold the correct data.
    * Storage: The task details should have the correct format after being extracted. So it is assumed to contains task name and date time by default, means it should be able to be split to the required length and hold the correct data.

Jiwei99 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+                || operation == Operation.EVENT) {
+                storage.addTask(command);
+            }
+            storage.saveTasksToFile();
```
Should include a default branch else here to detect errors. (Under Use the Default Branch section of Code Quality)

Jiwei99 commented on src/main/java/duke/controller/DukeController.java:
```diff
@@ -0,0 +1,167 @@
+            // The reason of following approach is because the size of the dukeDialogController
+            // will be returned only after it is added by the dialogContainer.
+            DukeDialogController preDukeDialogController = new DukeDialogController(content);
+            preDukeDialogController.heightProperty()
```
Can try to break this statement up to avoid complicated statements. Also, is it optimal to have nested try-catch block? (Avoid complicated expressions under Code Quality)

Jiwei99 commented on src/main/java/duke/controller/DukeController.java:
```diff
@@ -0,0 +1,167 @@
+            || operation == Operation.DEADLINE
+            || operation == Operation.EVENT) {
+            storage.addTask(input);
+        }
```
Similarly, should have a else block for error handling

HolmesJJ committed:
Implement CodeQuality

Improving the Code Quality based on the following rule:
* Avoid Deep Nesting
* Avoid Magic Numbers and Avoid Magic Characters
* SLAP Hard
* Avoid Empty Catch Blocks
* Delete Dead Code

Update the indentation for wrapped lines to 8 spaces.

Jiwei99 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,180 @@
+     *
+     * @return Tasks size.
+     */
+    public int size() {
```
Names representing methods should be verbs, i.e. getSize() (Under coding standard)

Jiwei99 reviewed (commented):
Generally LGTM, some minor comments.

Chesterwongz reviewed (commented):
LGTM! I think you really went above and beyond the requirements! Good Job!

HolmesJJ committed:
Implement more OOP to improve Code Quality

Moving command operations logic from Storage class to different Command classes.

Creating an abstract command class and extracting more classes that inherits from it to handle the command from user input.
* AddCommand contains add operation for Todo, Deadline and Event task.
* ClearCommand contains clear all tasks operation.
* DeleteCommand contains delete one task operation.
* DoneCommand contains the operation of marking task as done.
* ExitCommand contains operation of showing exit message.
* FindCommand contains find matching tasks operation.
* ListCommand contains list all tasks operation.

Creating a Parser class to return the respective Command object.

HolmesJJ committed:
Create ci.yml

HolmesJJ committed:
Move test.txt

Move the location of test.txt from /ip/src/test/data/ to /ip/data.

HolmesJJ committed:
Implement Code Quality

Let's
* rename size() to getSize() since names representing methods should be verbs (Under coding standard).
* update comment of parse to show the case of returning empty.
* avoid Deep Nesting by breaking the addDukeDialog and addUserDialog methods.

HolmesJJ committed:
Merge pull request #2 Implement CodeQuality from branch-A-CodeQuality to Master

Implement CodeQuality

HolmesJJ committed:
Pull Master into branch-A-Assertions

# Conflicts:
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/task/TaskList.java

HolmesJJ committed:
Fix bug of merging PR

Some updating in the previous commit "d8189da9ebdcd2c0f648ce6966e03ba0f95b4544 [d8189da]" is missing after merging.

HolmesJJ committed:
Merge pull request #3 Implement Assertions from branch-A-Assertions to Master

Implement Assertions

HolmesJJ committed:
Implement Reminders

Adding thread pool to control the timer of the Reminder. The system will automatically remind the user how many tasks have not been completed in the coming few hours.
Adding new command "coming" to show the coming tasks.
Adding notification on UI to show how many how many tasks have not been completed in the coming few hours.

HolmesJJ committed:
Merge pull request #4 Implement Reminders from branch-B-Reminders to Master

Implement Reminders

HolmesJJ committed:
Pull Master into branch-A-CI

HolmesJJ committed:
Fix bat and bash bugs for passing testing

HolmesJJ committed:
Pull Master into branch-A-CI

HolmesJJ committed:
Add dos2unix command for Linux and MacOS

HolmesJJ committed:
Update 12-hour clock to uppercase

The output of DateTimeFormatter on MacOS is different from Linux. MacOS is lowercase, but Linux is uppercase.
So converting the DateTimeFormatter to lowercase manually.

HolmesJJ committed:
Pull Master into branch-A-CI

HolmesJJ committed:
Merge pull request #1 Create ci.yml from branch-A-CI to Master

Create ci.yml

HolmesJJ committed:
Implement Better Gui

HolmesJJ committed:
Add reused comments

HolmesJJ committed:
Update README and Ui

HolmesJJ committed:
Set theme jekyll-theme-cayman

HolmesJJ committed:
Update comments to follow the coding standard

HolmesJJ committed:
Update UserGuide

HolmesJJ committed:
Update comments

HolmesJJ committed:
Add CI and codecov to README

HolmesJJ committed:
Add Saving the data and Command Summary to README.md

HolmesJJ committed:
Update exclamation mark
